### ISSUE TITLE
UCTNode::create_children - wrong lock-handling?

### ISSUE THREAD

**Issue by snoewchen** [glinscott/leela-chess#537]
Between acquiring an releasing the lock (in UCTNode::create_children) there are 2 checks if "somebody beat us to it". If the check fails (e.g. somebody beat us to it) the function immediately returns - but then the lock never gets released.

I guess this is not a problem because the children are created only once and every node has it's own mutex object - but on the other hand - why bother releasing the lock in the first place?

Wait ... if there are 2 threads waiting for the critical section (a third thread won) - the 2. thread will never resume because the lock never gets released...

(And sorry if I misunderstood the source code completely - C++ is not my first language...)

**Comment by killerducky** [comment]
I'm not a C++ expert either, but LOCK is a macro that creates a locally scoped variable `lock`. You can manually do lock.unlock() to release the lock before the variable goes out of scope. But if you do nothing and the function just ends, the variable goes out of scope. The destructor then automatically calls unlock for you.

**Comment by snoewchen** [comment]
Hi @killerducky, thanks for the clarification - you are right - I misunderstood the source code completely (on the plus side: I have now a working Code::Blocks debug-environment)
